pinot-core/src/main/java/org/apache/pinot/core/transport/InstanceRequestHandler.java

```diff
@@ -321,6 +322,10 @@ private void sendResponse(ChannelHandlerContext ctx, String tableNameWithType, l
         LOGGER.info("Slow query: request handler processing time: {}, send response latency: {}, total time to handle "
             + "request: {}", queryProcessingTimeMs, sendResponseLatencyMs, totalQueryTimeMs);
       }
+      if (serializedDataTable.length > LARGE_RESPONSE_SIZE_THRESHOLD_BYTES) {
+        LOGGER.warn("Large query: response size in bytes: {}, table name {}",
+            serializedDataTable.length, tableNameWithType);
```

Comment:
I don't think this warning serves any purpose. 

Can we not just kill the query on the server when we detect this ?

Reply:
we could do that even before allocating the buffer if we add this limit in `LengthFieldBasedFrameDecoder`.

I think we could just add a PinotProperty that defines the max dataTable size and set that to something like 100-500 by default. With that single change, we could avoid all these issues. That may be problematic in cases where a query requires a very large request from a server, but that is not a normal kind of queries in a OLAP system and it could be changed at startup time if needed

Reply:
I think instead of a warning, emitting a metric may be much better to help us plan future enhancements IMO.

- Understanding why response is so large and what can be done in terms of optimizing it

- If the server knows that the fan out of query is huge then there is a chance of having other servers' response size huge as well (and therefore a chance for causing Direct Memory OOM on broker). In that case, killing it on the server before responding is better. 

> Can we not just kill the query on the server when we detect this ?


I take this back. Simply killing upon detecting response on server is likely to lead to lot of pre-mature / unnecessary queries getting killed. So may be we make servers aware of fan-out and then decide to kill.

Reply:
> I think instead of a warning, emitting a metric may be much better to help us plan future enhancements IMO.

Yes, this is why I'd like to have a histogram metric for the size of message sent from the Server to a Broker. Then we can figure out a reliable max size over a period of time where this OOM does not happen and use that for setting a max size threshold.

My opinion here is that there are several ways that we can have a more elegant solution but dealing with the Direct Buffer OOM is still a fundamental requirement we should have. And I think the right first step is to have a way to deal with that DM OOM and collect additional data about what we are sending between nodes, then use that data to understand how how to tune the more nuanced but more elegant next steps.

Reply:
Yes let's get together on how to improve this in future. I think @jasperjiaguo  has already sync'd up with some of you offline on slack on that.

> I think we could just add a PinotProperty that defines the max dataTable size and set that to something like 100-500 by default.

I think this will work but my concern is that it may lead to a lot of pre-mature query killing and will introduce another config. Fan out is also a factor that we should consider. 

Let's all do some brainstorming separately.

Reply:
@siddharthteotia yes, the risk with the message size threshold is that if we are not very careful in tuning we'll start killing previously fine queries. Hence the need for a histogram metric to get an idea of the size of the messages. Furthermore, it will heavily dependent on cluster hardware and configuration. It requires a lot of thought in terms of both how it's implemented, what the defaults are (or if it's off by default), and how we tune it in such a way that we don't break functional queries.

The other key point is: we still have to deal with this direct buffer OOM issue even with the message size threshold. Remember, the core issue here is that the Broker becomes a bad actor in the cluster and starts accepting queries and then, essentially, throwing them out. We have to address _that_ issue before we worry about more nuanced ways of reducing the likelihood of an OOM happening.

My proposal is that we do a multipart fix: 
1. we either use this fix or we treat an OOM as a fault and shut the Broker down (so that it is removed from the cluster asap and ceases to be a bad actor), 
2. once that's out, we add in metrics to collect about message properties, 
3. we add a message size threshold feature combined with a robust guide for enabling and tuning that will greatly reduce the risk of the OOM happening.

Reply:
+1 for shutting down the Broker.

Reply:
> yes, the risk with the message size threshold is that if we are not very careful in tuning we'll start killing previously fine queries. Hence the need for a histogram metric to get an idea of the size of the messages.

We have that metric. See `BrokerMeter.NETTY_CONNECTION_BYTES_SENT`, `BrokerMeter.NETTY_CONNECTION_BYTES_RECEIVED` and similar meters in servers. Also, the PR I linked adds see `ServerGauge.NETTY_POOLED_USED_DIRECT_MEMORY` and `BrokerGauge.NETTY_POOLED_USED_DIRECT_MEMORY` (but the once we already have are the one you asking for).

Also, we have other similar restrictions (like the implicit limit we have in queries) and we didn't make a super-large study to know the exact number. It is enough to just add heuristic number and let the users change them with a config.